### PR TITLE
Streamlining typechecking of implicits solutions

### LINKED ISSUE #2229: mapply fails to rewrite when the goal uses the Ghost effect
When using `mapply` to rewrite the goal, for example to apply the theorem `forall x. id x == x`, it fails when `x`:
- uses the `Ghost` effect (see `test_ghost` below)
- uses the `GTot` effect inside a `Pure` effect (see `test_pure_gtot`)

However, it has no problem when `x` uses:
- `GTot` (see `test_gtot`)
- `Pure` (see `test_pure`)
- `Pure` inside a `GTot` (see `test_gtot_pure`)

```fstar
open FStar.Tactics

(* A helper function to rewrite things *)
let rw (t:term): Tac unit =
  pointwise (fun () ->
    try (mapply t)
    with _ -> trefl ()
  )

assume val id: int -> int
assume val id_def: x:int -> Lemma
  (id x == x)

[@expect_failure]
let test_ghost (x:int) (f:int -> Ghost (int) (requires True) (ensures fun _ -> True)): unit =
  assert (id (f x) == f x) by (
    rw(`id_def);
    trefl();
    qed()
  )

[@expect_failure]
let test_pure_gtot (x:int) (f:int -> Pure (int) (requires True) (ensures fun _ -> True)) (g:int -> GTot int): unit =
  assert (id (f (g x)) == f (g x)) by (
    rw(`id_def);
    trefl();
    qed()
  )

let test_gtot(x:int) (f:int -> GTot int): unit =
  assert (id (f x) == f x) by (
    rw(`id_def);
    trefl();
    qed()
  )

let test_pure (x:int) (f:int -> Pure (int) (requires True) (ensures fun _ -> True)): unit =
  assert (id (f x) == f x) by (
    rw(`id_def);
    trefl();
    qed()
  )

let test_gtot_pure (x:int) (f:int -> GTot int) (g:int -> Pure (int) (requires True) (ensures fun _ -> True)): unit =
  assert (id (f (g x)) == f (g x)) by (
    rw(`id_def);
    trefl();
    qed()
  )
```

### PR DESCRIPTION
**Pull request by aseemr**
## Using fast path by default for checking implicits

This PR streamlines the typechecking of implicits solutions by using fast path (i.e. faster computation of types when we know that the term is well-typed).

### Existing behavior

Suppose `?u:k` is a unification variable introduced at type `k`, and using higher-order unification, F* computes a solution `?u <- t`. Since unification in F* is integrated with subtyping and refinement types, F* has to make sure that `t:k`. For that, there is a loop at the end in the unifier that iterates over all the implicits and their solutions and checks\
 that `t:k` for each of them.

The typechecking `t:k` has two implementations in F*: one that uses the normal typechecker flow and one that optimizes typechecking using the knowledge that `t` is already well-typed _at some type_. For example, say the solution `t` is `f e`. While the normal typechecker flow would typecheck `f`, then `e`, and then bind their computation types etc., the optimized path (i.e. the fast path) would not re-typecheck `e`. It would compute the type of `f` (using fast path), and then substitute `e` in it for the bound variable to get the final type.

F* unifier by-default uses the normal typechecker flow, the option of fast path is hidden under an option that's unset by-default. When this loop of typechecking implicits is called from the tactics engine, there it defaults to fast path.

Using this _slow_ path for implicits has some disadvantages:

* First, it means that typechecking implicit solutions may generate more implicits, which will again need to be resolved. We have seen this kind of issues with tactics implicits.

* Second, it maintains two paths in the typechecker for checking implicits solutions.

* Finally, there may be some performance overhead associated with running the full typechecker again.

This PR streamlines the implicits checking by always using the fast path first, defaulting to full typechecker if it fails.

### Switching on fast path

When trying to switch on the fast path by default, we discovered that re-typechecking implicits solutions served another purpose: resolve some unification variables (especially universe ones) that remained unresolved until now. A small example is as follows:

```
type t1 (#a:Type u#a) (x:a) : Type0
type t2 (#a:Type u#a) (x:a) : Type0 = t1 x

```

To typecheck the body of `t2`, we have the following problem:

```
a:Type u#a, x:a |- t1 x : ?
```

When we typecheck `t1`, we instantiate its universe with say `?uv`. Then we introduce an implicit `?u` for the type argument to `t1`, s.t. `?u:Type ?uv`. Moving forward, for the application of `x`, we generate a subtyping constraint `a <: ?u`. Ultimately, we will solve `?u <- a`. But note that `?uv` remains hanging. In the current flow, it will get resolved w\
hen `a:Type u#a` is typechecked at `Type ?uv`.

### First cut at using fast path

A first cut solution then is as follows: for checking `?u <- t`, where `?u:k`:

* If either `t` or `k` have unresolved uvars go to slow path
* Compute the type of `t` using fast path
* If could not compute the type of `t` on fast path or the computed type has unresolved unification variables, go to slow path
* Check `k' <: k` where `k'` is the computed type of `t`

### Solution types optimization

Going back to the example:

```
type t1 (#a:Type u#a) (x:a) : Type0
type t2 (#a:Type u#a) (x:a) : Type0 = t1 x

```

We would be unable to use fast path for `?u` since its type would have an unresolved universe unification variable. To handle some of these cases, we implement an optimization in the unifier as follows:

Whenever the unifier sets a solution `?u <- t`, we note `(k1, k2)` where `t:k1` and `?u:k2` in the unifier state. In the end, we have a list of these `(k1, k2)` tuples. Just before we loop over the implicits to typecheck their solutions, we go over this list, and for cases when exactly one of `k1` or `k2` have unresolved uvars, we solve the problem `k1 <: k2`\
.

This helps us resolve some of the unification variables, allowing us to go to the fast path for implicits checking later on.

### Impact of the first cut fast path implementation

In `ulib`, overall we were checking 89159 implicits, all using slow path.

With the first cut solution, no optimization, we were able to use fast path for 60959 of those, leaving 28200 to slow path.

For HACL*, we were able to use fast path for 511122 implicits, leaving 53777 to slow path (earlier all of them were on the slow path).

### Impact of the optimization

With the solution types optimization, for `ulib` we were able to handle additional 11714 cases in fast path, thus finally using 72673 on fast path and 16486 on slow path.

For HACL*, we added 43495 cases more, leaving only 10282 to slow path after optimization.

### Impact on the running time

Impact on the verification time was quite modest. Since the typechecking for majority of implicits happens in phase 1 in the lax mode, it already was not a lot. In an everest run for example, we were spending 315365 ms in typechecking implicits. With fast path, this reduced to 21470 ms.

### Other items in the PR

* Handling more cases in fast path type and universe computation
* Better names for callbacks in env and `TcTerm` functions
* Fixing unifier to instantiate universes in `Pat_cons` when the `match` heuristic kicks in
* When combining the refinements in meet/join in the unifier, `squash` the formulas if needed, since they could be in some universe other than zero, in which case directly applying `l_or` or `l_and` constraints their universe to zero. Adding a `squash` avoids it

### Code pointers

* `TcTerm.fs:tc_check_tot_or_gtot_term_maybe_fastpath`: implements the fast path typechecking with fallback.
* `TcTerm.fs:typeof_tot_or_gtot_term_fastpath`: handling more cases, e.g. `Tm_app` when arguments are more, `Tm_abs` when residual type is not set, etc.
* `TcTerm.fs:universe_of_well_typed_term`: handling more cases, e.g. `eqtype` would fail earlier
* `Rel.fs`: addition of a new compartment of type `list (typ & typ)` in the guard and worklist (to implement the optimization).
* `PatternUtils.fs`: instantiating universes in `Pat_cons`
* `Env.fsi`: typechecker callbacks names

### Remaining issues

* When the fast path is invoked from tactics, the code remains as is. Using the same scheme as `Rel` for tactics is too slow, but always using fast path results in unresolved cases (in `FStar` examples). Left it as is for now, should revisit.

* There are several cases in the typechecker where we could call fast path (e.g. normalizer, extraction), have made a note of them.

* Can we reach a point when we don't rely on implicits solution typechecking to resolve unification variables?

Finally it also fixes #2229 (that's where it all started from a few days back).

### CODE REVIEW

File: src/tactics/FStar.Tactics.Basic.fs [hunk 1]
```diff
@@ -845,8 +845,9 @@ let t_apply_lemma (noinst:bool) (noinst_lhs:bool)
                   // NS:05/25: protecting it under this option,
                   //           since it causes a regression in examples/vale/*Math_i.fst
                   // GM: Made it the default, but setting must_total to true
-                  FStar.TypeChecker.TcTerm.check_type_of_well_typed_term'
-                            true env term ctx_uvar.ctx_uvar_typ
+                  // AR:03/17: These are lemma arguments, so we don't need to insist on must_total
+                  FStar.TypeChecker.TcTerm.tc_check_tot_or_gtot_term_maybe_fastpath
+                            env term ctx_uvar.ctx_uvar_typ false true
```
Comment (aseemr) [inline src/tactics/FStar.Tactics.Basic.fs:850]: cc @nikswamy , setting `must_tot` to `false` here. These are lemma arguments, so ghost is fine here.

File: src/typechecker/FStar.TypeChecker.TcTerm.fs [hunk 2]
```diff
@@ -3901,57 +3919,149 @@ let rec type_of_well_typed_term (env:env) (t:term) : option<typ> =
...
+    aux true k)
+
+(*
+ * Precondition: G |- t : Tot _ or G |- t : GTot _
+ *   Meaning, even if t is PURE or GHOST, its wp has been accounted for already,
+ *   which is the case for terms in the unifier
+ *
+ * It returns either PURE or GHOST (or None if fast path fails)
+ *)
+let rec effectof_tot_or_gtot_term_fastpath (env:env) (t:term) : option<lident> =
```
Comment (aseemr) [inline src/typechecker/FStar.TypeChecker.TcTerm.fs:3950]: cc @nikswamy , implementation of `effectof_tot_or_gtot_term_fastpath`.

File: src/typechecker/FStar.TypeChecker.TcTerm.fs [hunk 3]
```diff
@@ -3901,57 +3919,149 @@ let rec type_of_well_typed_term (env:env) (t:term) : option<typ> =
...
+         match Rel.subtype_nosmt env k' k with
+         | None -> slow_path "subtype_nosmt returned None"
+         | Some g -> g
+  else
+    (*
+     * AR: We sometimes rely on typechecking the implicit solutions to
+     *       resolve some unification variables, especially universe uvars
+     *     So, if there are any uvars in t, k, or the computed type, default to slow path
+     *)
+    if not (continue_fast_path t) then slow_path "t has uvars"
```
Comment (aseemr) [inline src/typechecker/FStar.TypeChecker.TcTerm.fs:4044]: cc @nikswamy , fast path algorithm.

File: src/typechecker/FStar.TypeChecker.TcTerm.fs [hunk 4]
```diff
@@ -3849,50 +3869,48 @@ let rec type_of_well_typed_term (env:env) (t:term) : option<typ> =
...
                let bs, c = SS.open_comp bs (S.mk_Total t) in
-               Some (bs, U.comp_result c)
-          else if n_args = n_bs
-          then let bs, c = SS.open_comp bs c in
-               if U.is_tot_or_gtot_comp c
-               then Some (bs, U.comp_result c)
-               else None
-          else None
+               Some (bs, args, U.comp_result c, [])
+          else let bs, c = SS.open_comp bs c in
```
Comment (aseemr) [inline src/typechecker/FStar.TypeChecker.TcTerm.fs:3901]: cc @nikswamy , handling case when args are more than binders.

File: src/typechecker/FStar.TypeChecker.Rel.fs [hunk 5]
```diff
@@ -852,9 +867,22 @@ let solve_prob' resolve_ok prob logical_guard uvis wl =
              wl
     in
     commit uvis;
-    {wl with ctr=wl.ctr + 1}
-
-let extend_solution pid sol wl =
+    //record type of the uvar and its solution in the worklist
```
Comment (aseemr) [inline src/typechecker/FStar.TypeChecker.Rel.fs:870]: cc @nikswamy , when assigning `?u <- t`, we record `(k1, k2)`, where `t:k1` and `?u:k2`, in the worklist.

File: src/typechecker/FStar.TypeChecker.Rel.fs [hunk 6]
```diff
@@ -4196,18 +4238,32 @@ let resolve_implicits' env is_tac g =
           end
         end
   in
+
+  (*
```
Comment (aseemr) [inline src/typechecker/FStar.TypeChecker.Rel.fs:4242]: cc @nikswamy , before starting fixpoint, iterate over the `solution_types` compartment.

File: src/typechecker/FStar.TypeChecker.Rel.fs [hunk 7]
```diff
@@ -1869,6 +1869,19 @@ and solve_rigid_flex_or_flex_rigid_subtyping
...
+                   *       meaning that if we apply op (l_and or l_or) directly, we may be
+                   *       unifying the universe of phi to zero, leading to errors
+                   *)
+                  let apply_op op phi1 phi2 =
+                    let squash phi =
+                      match env.universeof_fastpath env phi with
+                      | Some U_zero -> phi
+                      | Some u -> U.mk_squash u phi
+                      | None -> U.mk_squash (Env.new_u_univ ()) phi in
+                    op (squash phi1) (squash phi2) in
```
Comment (aseemr) [inline src/typechecker/FStar.TypeChecker.Rel.fs:1884]: cc @nikswamy , using `squash` when combining the refinements.

File: src/typechecker/FStar.TypeChecker.Env.fs [hunk 8]
```diff
@@ -175,10 +175,11 @@ and env = {
   nosynth        :bool;                         (* don't run synth tactics *)
   uvar_subtyping :bool;
   tc_term        :env -> term -> term*lcomp*guard_t; (* a callback to the type-checker; g |- e : M t wp *)
-  type_of        :env -> term -> term*typ*guard_t;   (* a callback to the type-checker; g |- e : Tot t *)
-  type_of_well_typed :env -> term -> typ;          (* a callback to the type-checker; falls back on type_of if the term is not well-typed *)
+  typeof_tot_or_gtot_term:env -> term -> bool -> term*typ*guard_t;   (* a callback to the type-checker; g |- e : Tot t *)
+  typeof_tot_or_gtot_term_fastpath  :env -> term -> option<typ>;       (* a callback to the type-checker, uses fast path *)
```
Comment (nikswamy) [inline src/typechecker/FStar.TypeChecker.Env.fs:179]: From our discussions, I was expecting this to check and return whether the term is PURE or GHOST

File: src/typechecker/FStar.TypeChecker.Env.fs [hunk 9]
```diff
@@ -175,10 +175,11 @@ and env = {
...
   universe_of    :env -> term -> universe;           (* a callback to the type-checker; g |- e : Tot (Type u) *)
-  check_type_of  :bool -> env -> term -> typ -> guard_t;
+  tc_check_tot_or_gtot_term_maybe_fastpath :env -> term -> typ -> bool -> bool -> guard_t;
+  universeof_fastpath    :env -> term -> option<universe>;
```
Comment (nikswamy) [inline src/typechecker/FStar.TypeChecker.Env.fs:182]: Would be good to update the comments here. And to indicate there why we are exposing three versions of the function in the interface, i.e., typeof_, typeof_..._fastpath, tc_check_..._maybe_fastpath

Comment (nikswamy) [reply src/typechecker/FStar.TypeChecker.Env.fs:182]: likewise for universeof

File: src/typechecker/FStar.TypeChecker.Env.fsi [hunk 10]
```diff
@@ -441,3 +444,5 @@ val pure_precondition_for_trivial_post : env -> universe -> typ -> typ -> Range.
 for either not a recursive let, or one that does not need the totality
 check. *)
 val get_letrec_arity : env -> lbname -> option<int>
+
+val tc_tot_or_gtot_term_maybe_fastpath : env -> term -> typ
```
Comment (nikswamy) [inline src/typechecker/FStar.TypeChecker.Env.fsi:448]: And here's a fourth variant?

File: src/typechecker/FStar.TypeChecker.Normalize.fs [hunk 11]
```diff
@@ -2911,10 +2911,10 @@ let eta_expand (env:Env.env) (t:term) : term =
         let formals, _tres = U.arrow_formals (SS.subst' s u.ctx_uvar_typ) in
         if List.length formals = List.length args
         then t
-        else let _, ty, _ = env.type_of ({env with lax=true; use_bv_sorts=true; expected_typ=None}) t in
+        else let _, ty, _ = env.typeof_tot_or_gtot_term ({env with lax=true; use_bv_sorts=true; expected_typ=None}) t true in
```
Comment (nikswamy) [inline src/typechecker/FStar.TypeChecker.Normalize.fs:2914]: I wonder if it would be more readable if the interface provided typeof_tot_term and type_of_gtot_term, rather than a single function whose behavior is toggled with a flag. The `true` at the end is a bit opaque to read

Comment (aseemr) [reply src/typechecker/FStar.TypeChecker.Normalize.fs:2914]: Sounds good to me, will change.

File: src/typechecker/FStar.TypeChecker.Env.fs (hunk 9 above)
Comment (aseemr) [reply src/typechecker/FStar.TypeChecker.Env.fs:182]: I will add a comment. But here is my reading of it:

* The `tc_` functions have a return type of the form `env -> term -> term & lcomp & guard_t`
* The `tc_check_` functions in addition take an expected type as argument: `env -> term -> typ -> term & lcomp & guard_t`
* The `typeof_` functions return only a type not complete lcomp: `env -> term -> term & typ & guard_t`
* Then there are `_fastpath` variants, which I added, that take the fastpath (and return no guard for example).
* The `maybe_fastpath` variant first tries the fastpath, and if it fails, tries the slow path.

File: src/typechecker/FStar.TypeChecker.DeferredImplicits.fs [hunk 12]
```diff
@@ -134,7 +134,7 @@ let solve_deferred_to_tactic_goals env g =
             if is_flex tp.lhs then tp.lhs
             else tp.rhs
           in
-          env.type_of env_lax t
+          env.typeof_tot_or_gtot_term env_lax t true  //AR: TODO: can we call type_of_well_typed?
```
Comment (nikswamy) [inline src/typechecker/FStar.TypeChecker.DeferredImplicits.fs:137]: What is the TODO about?

File: src/typechecker/FStar.TypeChecker.Rel.fs [hunk 13]
```diff
@@ -1476,7 +1504,7 @@ let rec really_solve_universe_eq pid_orig wl u1 u2 =
                                         (Print.univ_to_string u2))
 
         | U_name x, U_name y ->
-          if (string_of_id x) = (string_of_id y)
+          if (string_of_id x) = (string_of_id y) || Options.ml_ish ()
```
Comment (nikswamy) [inline src/typechecker/FStar.TypeChecker.Rel.fs:1507]: this is slightly surprising? Why the ml_ish here?

File: src/typechecker/FStar.TypeChecker.Rel.fs (hunk 7 above)
Comment (nikswamy) [reply src/typechecker/FStar.TypeChecker.Rel.fs:1884]: nice catch

File: src/typechecker/FStar.TypeChecker.Rel.fs [hunk 14]
```diff
@@ -2918,7 +2966,7 @@ and solve_t' (env:Env.env) (problem:tprob) (wl:worklist) : solution =
         let force_eta t =
             if is_abs t then t
             else begin
-                let _, ty, _ = env.type_of ({env with lax=true; use_bv_sorts=true; expected_typ=None}) t in
+                let _, ty, _ = env.typeof_tot_or_gtot_term ({env with lax=true; use_bv_sorts=true; expected_typ=None}) t true in  //AR: TODO: type_of_well_typed?
```
Comment (nikswamy) [inline src/typechecker/FStar.TypeChecker.Rel.fs:2969]: Would be nice to use type_of_well_typed and remove the lax=true ... but that can be a separate PR

File: src/typechecker/FStar.TypeChecker.Rel.fs (hunk 13 above)
Comment (aseemr) [reply src/typechecker/FStar.TypeChecker.Rel.fs:1507]: Thanks for catching it. At some point bootstrapping was failing so I added it, removed now.
